apps/microtvm/arduino/template_project/microtvm_api_server.py

```diff
@@ -0,0 +1,423 @@
+        return f"{o['package']}:{o['architecture']}:{o['board']}"
+
+    def build(self, options):
+        BUILD_DIR.mkdir()
+        print(BUILD_DIR)
+
+        compile_cmd = [
+            options["arduino_cli_cmd"],
+            "compile",
+            "./project/",
+            "--fqbn",
+            self._get_fqbn(options),
+            "--build-path",
+            BUILD_DIR.resolve(),
+        ]
+
+        if options.get("verbose"):
+            compile_cmd.append("--verbose")
+
+        # Specify project to compile
+        output = subprocess.check_call(compile_cmd)
+        assert output == 0
+
+    # We run the command `arduino-cli board list`, which produces
+    # outputs of the form:
+    """
+    Port         Type              Board Name FQBN                          Core
+    /dev/ttyS4   Serial Port       Unknown
+    /dev/ttyUSB0 Serial Port (USB) Spresense  SPRESENSE:spresense:spresense SPRESENSE:spresense
+    """
+
+    def _auto_detect_port(self, options):
+        list_cmd = [options["arduino_cli_cmd"], "board", "list"]
+        list_cmd_output = subprocess.check_output(list_cmd).decode("utf-8")
+        # Remove header and new lines at bottom
+        port_options = list_cmd_output.split("\n")[1:-2]
+
+        # Select the first compatible board
+        fqbn = self._get_fqbn(options)
+        for port_option in port_options:
+            if fqbn in port_option:
+                return port_option.split(" ")[0]
+
+        # If no compatible boards, raise an error
+        raise BoardAutodetectFailed()
+
+    def _get_arduino_port(self, options):
+        if not self._port:
+            if "port" in options and options["port"]:
+                self._port = options["port"]
+            else:
+                self._port = self._auto_detect_port(options)
+
+        return self._port
+
+    def flash(self, options):
+        port = self._get_arduino_port(options)
+
+        upload_cmd = [
+            options["arduino_cli_cmd"],
+            "upload",
+            "./project",
+            "--fqbn",
+            self._get_fqbn(options),
+            "--input-dir",
+            BUILD_DIR.resolve(),
+            "--port",
+            port,
+        ]
+
+        if options.get("verbose"):
+            upload_cmd.append("--verbose")
+
+        output = subprocess.check_call(upload_cmd)
+
+        if output == 2:
+            raise InvalidPortException()
+        elif output > 0:
+            raise SketchUploadException()
+
+    def open_transport(self, options):
+        # Zephyr example doesn't throw an error in this case
+        if self._serial is not None:
+            return
+
+        port = self._get_arduino_port(options)
+
+        # Wait for port to become available
+        for attempts in range(10):
+            if any(serial.tools.list_ports.grep(port)):
+                break
+            time.sleep(0.5)
+
+        # TODO figure out why RPC serial communication times out 90% (not 100%)
```

Comment:
can you include stacktrace as a GH issue and link here?

Reply:
Created #8730 to address it, and removed the comment.